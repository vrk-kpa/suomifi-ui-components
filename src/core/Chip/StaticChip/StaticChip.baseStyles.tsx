import { SuomifiTheme } from '../../theme';
import { css } from 'styled-components';
import { baseChipBaseStyles } from '../BaseChip/BaseChip.baseStyles';
import { MarginProps, buildSpacingCSS } from '../../theme/utils/spacing';

export const staticChipBaseStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  ${baseChipBaseStyles(theme)}
  ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)}
`;
