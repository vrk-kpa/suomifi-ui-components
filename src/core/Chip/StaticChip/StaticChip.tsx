import React, { Component, forwardRef } from 'react';
import classnames from 'classnames';
import { default as styled } from 'styled-components';
import { HtmlSpan, HtmlSpanProps } from '../../../reset';
import {
  BaseChipProps,
  baseClassName,
  chipClassNames,
} from '../BaseChip/BaseChip';
import { staticChipBaseStyles } from './StaticChip.baseStyles';
import {
  SuomifiThemeProp,
  SuomifiThemeConsumer,
  SpacingConsumer,
} from '../../theme';
import {
  separateMarginProps,
  MarginProps,
  GlobalMarginProps,
} from '../../theme/utils/spacing';
import { filterDuplicateKeys } from '../../../utils/common/common';

export interface StaticChipProps
  extends BaseChipProps,
    MarginProps,
    Omit<HtmlSpanProps, 'children'> {
  /** Ref is forwarded to the span element. Alternative for React `ref` attribute. */
  forwardedRef?: React.Ref<HTMLSpanElement>;
}

class BaseChip extends Component<StaticChipProps> {
  render() {
    const { className, children, disabled = false, ...rest } = this.props;
    const [_marginProps, passProps] = separateMarginProps(rest);

    return (
      <HtmlSpan
        className={classnames(baseClassName, className, {
          [chipClassNames.disabled]: !!disabled,
        })}
        {...passProps}
        style={{ ...passProps?.style }}
      >
        <HtmlSpan className={chipClassNames.content}>{children}</HtmlSpan>
      </HtmlSpan>
    );
  }
}

const StyledChip = styled(
  ({
    theme,
    globalMargins,
    ...passProps
  }: StaticChipProps & SuomifiThemeProp & GlobalMarginProps) => (
    <BaseChip {...passProps} />
  ),
)`
  ${({ theme, globalMargins, ...rest }) => {
    const [marginProps, _passProps] = separateMarginProps(rest);
    const cleanedGlobalMargins = filterDuplicateKeys(
      globalMargins.staticChip,
      marginProps,
    );
    return staticChipBaseStyles(theme, cleanedGlobalMargins, marginProps);
  }}
`;

const StaticChip = forwardRef(
  (props: StaticChipProps, ref: React.Ref<HTMLSpanElement>) => (
    <SpacingConsumer>
      {({ margins }) => (
        <SuomifiThemeConsumer>
          {({ suomifiTheme }) => (
            <StyledChip
              theme={suomifiTheme}
              globalMargins={margins}
              forwardedRef={ref}
              {...props}
            />
          )}
        </SuomifiThemeConsumer>
      )}
    </SpacingConsumer>
  ),
);

StaticChip.displayName = 'StaticChip';
export { StaticChip };
