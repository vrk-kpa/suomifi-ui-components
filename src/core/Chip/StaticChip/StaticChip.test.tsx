import React from 'react';
import { render } from '@testing-library/react';
import { StaticChip } from './StaticChip';

describe('disabled', () => {
  const DisabledChip = <StaticChip disabled={true}>Testcontent</StaticChip>;

  it('has "--disabled"-class', () => {
    const { container } = render(DisabledChip);
    expect(container.firstChild).toHaveClass('fi-chip--disabled');
  });

  it('should match snapshot', () => {
    const { container } = render(DisabledChip);
    expect(container).toMatchSnapshot();
  });
});

describe('children', () => {
  const chipWithChildren = (
    <StaticChip>
      <div data-testid="test-div">Selection 1</div>
    </StaticChip>
  );

  it('has the given content', () => {
    const { getByTestId } = render(chipWithChildren);
    expect(getByTestId('test-div').textContent).toBe('Selection 1');
  });

  it('should match snapshot', () => {
    const { container } = render(chipWithChildren);
    expect(container).toMatchSnapshot();
  });
});

describe('classnames', () => {
  const customClassChip = (
    <StaticChip className="custom-class">Testcontent</StaticChip>
  );
  const staticChip = <StaticChip>Test selection</StaticChip>;

  it('contains base classname', () => {
    const { container } = render(customClassChip);
    expect(container.firstChild).toHaveClass('fi-chip');
  });

  it('contains custom classname', () => {
    const { container } = render(customClassChip);
    expect(container.firstChild).toHaveClass('custom-class');
  });

  it('should match snapshot', () => {
    const { container } = render(customClassChip);
    expect(container).toMatchSnapshot();
  });
  it('only has the clasees corresponding the variant', () => {
    const { container } = render(staticChip);
    expect(container.firstChild).not.toHaveClass('fi-chip--button');
  });

  describe('margin prop', () => {
    it('should have margin style from margin prop', () => {
      const { container } = render(<StaticChip margin="xs">Test</StaticChip>);
      expect(container.firstChild).toHaveStyle('margin: 10px');
    });

    it('should have margin prop style overwritten from style', () => {
      const { container } = render(
        <StaticChip margin="xs" style={{ margin: 2 }}>
          Test
        </StaticChip>,
      );
      expect(container.firstChild).toHaveAttribute('style', 'margin: 2px;');
    });
  });
});
