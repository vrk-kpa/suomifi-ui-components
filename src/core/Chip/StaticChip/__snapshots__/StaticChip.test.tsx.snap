// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`children should match snapshot 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c1 {
  color: hsl(0,0%,13%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
}

.c1.fi-chip {
  border-radius: 14px;
  padding: 2px 10px;
  color: hsl(0,0%,100%);
  background: hsl(212,63%,45%);
  max-height: 100%;
  display: inline-block;
}

.c1.fi-chip:focus {
  outline: 0;
  position: relative;
}

.c1.fi-chip:focus::after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
  border-radius: 16px;
}

.c1.fi-chip .fi-chip--content {
  display: block;
  word-break: break-word;
  overflow: hidden;
  line-height: 1.5em;
}

.c1.fi-chip--disabled.fi-chip {
  cursor: not-allowed;
  background: hsl(202,7%,67%);
}

.c1.fi-chip--disabled.fi-chip:hover,
.c1.fi-chip--disabled.fi-chip:active {
  background: hsl(202,7%,67%);
}

@media screen and (forced-colors:active) {
  .c1.fi-chip--disabled.fi-chip {
    color: GrayText;
  }
}

<div>
  <span
    class="c0 fi-chip c1"
  >
    <span
      class="c0 fi-chip--content"
    >
      <div
        data-testid="test-div"
      >
        Selection 1
      </div>
    </span>
  </span>
</div>
`;

exports[`classnames should match snapshot 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c1 {
  color: hsl(0,0%,13%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
}

.c1.fi-chip {
  border-radius: 14px;
  padding: 2px 10px;
  color: hsl(0,0%,100%);
  background: hsl(212,63%,45%);
  max-height: 100%;
  display: inline-block;
}

.c1.fi-chip:focus {
  outline: 0;
  position: relative;
}

.c1.fi-chip:focus::after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
  border-radius: 16px;
}

.c1.fi-chip .fi-chip--content {
  display: block;
  word-break: break-word;
  overflow: hidden;
  line-height: 1.5em;
}

.c1.fi-chip--disabled.fi-chip {
  cursor: not-allowed;
  background: hsl(202,7%,67%);
}

.c1.fi-chip--disabled.fi-chip:hover,
.c1.fi-chip--disabled.fi-chip:active {
  background: hsl(202,7%,67%);
}

@media screen and (forced-colors:active) {
  .c1.fi-chip--disabled.fi-chip {
    color: GrayText;
  }
}

<div>
  <span
    class="c0 fi-chip c1 custom-class"
  >
    <span
      class="c0 fi-chip--content"
    >
      Testcontent
    </span>
  </span>
</div>
`;

exports[`disabled should match snapshot 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c1 {
  color: hsl(0,0%,13%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
}

.c1.fi-chip {
  border-radius: 14px;
  padding: 2px 10px;
  color: hsl(0,0%,100%);
  background: hsl(212,63%,45%);
  max-height: 100%;
  display: inline-block;
}

.c1.fi-chip:focus {
  outline: 0;
  position: relative;
}

.c1.fi-chip:focus::after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
  border-radius: 16px;
}

.c1.fi-chip .fi-chip--content {
  display: block;
  word-break: break-word;
  overflow: hidden;
  line-height: 1.5em;
}

.c1.fi-chip--disabled.fi-chip {
  cursor: not-allowed;
  background: hsl(202,7%,67%);
}

.c1.fi-chip--disabled.fi-chip:hover,
.c1.fi-chip--disabled.fi-chip:active {
  background: hsl(202,7%,67%);
}

@media screen and (forced-colors:active) {
  .c1.fi-chip--disabled.fi-chip {
    color: GrayText;
  }
}

<div>
  <span
    class="c0 fi-chip c1 fi-chip--disabled"
  >
    <span
      class="c0 fi-chip--content"
    >
      Testcontent
    </span>
  </span>
</div>
`;
